File: VB/Reporting_how-to-create-a-label-report-at-runtime-t473792/LabelReportRuntime/CustomLabelReportBuilder.vb
Imports DevExpress.XtraReports.UI
Imports DevExpress.XtraReports.Wizards
Imports DevExpress.XtraPrinting
Imports System.Drawing.Printing
Imports System
Imports DevExpress.XtraReports.Wizards.Labels
Imports System.Linq
Imports System.Drawing
Imports DevExpress.XtraPrinting.Native

Namespace dxWinFormsSample
    Public Class CustomLabelReportBuilder
        Public Sub New()
        End Sub
        Public Function GenerateLabelReport(ByVal width As Single, ByVal height As Single, ByVal vPitch As Single, ByVal hPitch As Single, ByVal measurementUnit As System.Drawing.GraphicsUnit, ByVal bottomMargin As Single, ByVal topMargin As Single, ByVal leftMargin As Single, ByVal rightMargin As Single, ByVal paperKindID As Integer) As XtraReport
            Dim model As New CustomLabelReportModel()
            model.LabelWidth = width
            model.LabelHeight = height
            model.VerticalPitch = vPitch
            model.HorizontalPitch = hPitch
            model.MeasurementUnit = measurementUnit
            model.BottomMargin = bottomMargin
            model.LeftMargin = leftMargin
            model.TopMargin = topMargin
            model.RightMargin = rightMargin
            model.PaperKindID = paperKindID
            Return BuildLabelReport(model)
        End Function
        Public Function GenerateLabelReport(ByVal model As CustomLabelReportModel) As XtraReport
            Return BuildLabelReport(model)
        End Function
        Private Function BuildLabelReport(ByVal model As CustomLabelReportModel) As XtraReport
            Dim report = New XtraReport()
            Dim paperKindList As PaperKindList = InitPaperKindList(model, report)
            report.ReportUnit = If(model.MeasurementUnit = System.Drawing.GraphicsUnit.Millimeter, ReportUnit.TenthsOfAMillimeter, ReportUnit.HundredthsOfAnInch)
            report.PaperKind = paperKindList.PaperKind
            report.Landscape = paperKindList.Landscape
            report.RollPaper = False
            Dim labelDpi As Single = GraphicsDpi.UnitToDpi(model.MeasurementUnit)
            Dim top As Integer = CInt((XRConvert.Convert(model.TopMargin, labelDpi, report.Dpi)))
            Dim left As Integer = CInt((XRConvert.Convert(model.LeftMargin, labelDpi, report.Dpi)))
            Dim right As Integer = CInt((XRConvert.Convert(model.RightMargin, labelDpi, report.Dpi)))
            Dim bottom As Integer = CInt((XRConvert.Convert(model.BottomMargin, labelDpi, report.Dpi)))
            report.Margins = New Margins(left, right, top, bottom)
            Dim labelWidth As Single = XRConvert.Convert(model.LabelWidth, labelDpi, report.Dpi)
            Dim labelHeight As Single = XRConvert.Convert(model.LabelHeight, labelDpi, report.Dpi)
            Dim detail As New DetailBand()
            report.Bands.Add(detail)
            Dim panel As New XRPanel()
            panel.CanGrow = False
            detail.Controls.Add(panel)
            panel.WidthF = labelWidth
            panel.HeightF = labelHeight
            panel.Borders = BorderSide.All
            detail.HeightF = XRConvert.Convert(model.VerticalPitch, labelDpi, report.Dpi)

            Dim labelInfo = DevExpress.XtraReports.Wizards.Builder.LabelInfoFactory.Create(GetDefaultLabelReportMode(model))
            report.ReportPrintOptions.DetailCountOnEmptyDataSource = labelInfo.CalculateLabelCount(paperKindList.Size, paperKindList.UnitDpi)
            If labelInfo.MoreOneColumnOnPage(paperKindList.Size, paperKindList.UnitDpi) Then
                detail.MultiColumn.ColumnWidth = XRConvert.Convert(labelInfo.LabelWidth, labelDpi, report.Dpi)
                detail.MultiColumn.ColumnSpacing = XRConvert.Convert(labelInfo.HPitch - labelInfo.LabelWidth, labelDpi, report.Dpi)
                detail.MultiColumn.Layout = ColumnLayout.AcrossThenDown
            End If
            Return report
        End Function
        Private Function GetDefaultLabelReportMode(ByVal model As CustomLabelReportModel) As ReportModel
            Dim reportModel As New XtraReportModel()
            reportModel.ReportType = ReportType.Label
            Dim customLblInfo = New CustomLabelInformation()
            customLblInfo.Height = model.LabelHeight
            customLblInfo.Width = model.LabelWidth
            customLblInfo.TopMargin = model.TopMargin
            customLblInfo.BottomMargin = model.BottomMargin
            customLblInfo.LeftMargin = model.LeftMargin
            customLblInfo.RightMargin = model.RightMargin
            customLblInfo.VerticalPitch = model.VerticalPitch
            customLblInfo.HorizontalPitch = model.HorizontalPitch
            customLblInfo.PaperKindDataId = model.PaperKindID
            customLblInfo.Unit = model.MeasurementUnit
            reportModel.CustomLabelInformation = customLblInfo
            Return reportModel
        End Function
        Private Shared Function InitPaperKindList(ByVal model As CustomLabelReportModel, ByVal report As XtraReport) As PaperKindList
            Dim paperKindList = New PaperKindList(report.Dpi)
            Dim labelRepository = (New LabelProductRepositoryFactory()).Create()
            Dim paperKindItems = labelRepository.GetSortedPaperKinds().Select(AddressOf ConvertToPaperKindItem)
            For Each item In paperKindItems
                paperKindList.Add(item)
            Next item
            paperKindList.CurrentID = model.PaperKindID
            Return paperKindList
        End Function
        Private Shared Function ConvertToPaperKindItem(ByVal paperKindData As PaperKindData) As PaperKindItem
            Dim dpi As Single = DevExpress.XtraPrinting.GraphicsDpi.UnitToDpi(paperKindData.Unit)
            Dim paperKind = CType(paperKindData.EnumId, PaperKind)
            Dim size As SizeF = If(paperKind = System.Drawing.Printing.PaperKind.Custom, New SizeF(paperKindData.Width, paperKindData.Height), PageSizeInfo.GetPageSizeF(paperKind, dpi, PageSizeInfo.DefaultSize))
            Return New PaperKindItem(paperKindData.Name, size, paperKindData.Id, paperKind, dpi, paperKindData.IsRollPaper)
        End Function
    End Class
End Namespace

File: CS/Reporting_how-to-create-a-label-report-at-runtime-t473792/LabelReportRuntime/CustomLabelReportBuilder.cs
using DevExpress.XtraReports.UI;
using DevExpress.XtraReports.Wizards;
using DevExpress.XtraPrinting;
using System.Drawing.Printing;
using System;
using DevExpress.XtraReports.Wizards.Labels;
using System.Linq;
using System.Drawing;
using DevExpress.XtraPrinting.Native;

namespace dxWinFormsSample {
    public class CustomLabelReportBuilder  {
        public CustomLabelReportBuilder() { }
        public XtraReport GenerateLabelReport(float width, float height, float vPitch, float hPitch, System.Drawing.GraphicsUnit measurementUnit, float bottomMargin, float topMargin, float leftMargin, float rightMargin, int paperKindID) {
            CustomLabelReportModel model = new CustomLabelReportModel();
            model.LabelWidth = width;
            model.LabelHeight = height;
            model.VerticalPitch = vPitch;
            model.HorizontalPitch = hPitch;
            model.MeasurementUnit = measurementUnit;
            model.BottomMargin = bottomMargin;
            model.LeftMargin = leftMargin;
            model.TopMargin = topMargin;
            model.RightMargin = rightMargin;
            model.PaperKindID = paperKindID;
            return BuildLabelReport(model);
        }
        public XtraReport GenerateLabelReport(CustomLabelReportModel model) {
            return BuildLabelReport(model);
        }        
        private XtraReport BuildLabelReport(CustomLabelReportModel model) {
            var report = new XtraReport();
            PaperKindList paperKindList = InitPaperKindList(model, report);
            report.ReportUnit = model.MeasurementUnit == System.Drawing.GraphicsUnit.Millimeter ? ReportUnit.TenthsOfAMillimeter : ReportUnit.HundredthsOfAnInch;
            report.PaperKind = paperKindList.PaperKind;
            report.Landscape = paperKindList.Landscape;
            report.RollPaper = false;
            float labelDpi = GraphicsDpi.UnitToDpi(model.MeasurementUnit);
            int top = (int)XRConvert.Convert(model.TopMargin, labelDpi, report.Dpi);
            int left = (int)XRConvert.Convert(model.LeftMargin, labelDpi, report.Dpi);
            int right = (int)XRConvert.Convert(model.RightMargin, labelDpi, report.Dpi);
            int bottom = (int)XRConvert.Convert(model.BottomMargin, labelDpi, report.Dpi);
            report.Margins = new Margins(left, right, top, bottom);
            float labelWidth = XRConvert.Convert(model.LabelWidth, labelDpi, report.Dpi);
            float labelHeight = XRConvert.Convert(model.LabelHeight, labelDpi, report.Dpi);
            DetailBand detail = new DetailBand();
            report.Bands.Add(detail);
            XRPanel panel = new XRPanel();
            panel.CanGrow = false;
            detail.Controls.Add(panel);
            panel.WidthF = labelWidth;
            panel.HeightF = labelHeight;
            panel.Borders = BorderSide.All;
            detail.HeightF = XRConvert.Convert(model.VerticalPitch, labelDpi, report.Dpi);

            var labelInfo = DevExpress.XtraReports.Wizards.Builder.ReportInfoFactory.CreateLabelInfo(GetDefaultLabelReportMode(model));
            report.ReportPrintOptions.DetailCountOnEmptyDataSource = labelInfo.CalculateLabelCount(paperKindList.Size, paperKindList.UnitDpi);
            if (labelInfo.MoreOneColumnOnPage(paperKindList.Size, paperKindList.UnitDpi)) {
                detail.MultiColumn.ColumnWidth = XRConvert.Convert(labelInfo.LabelWidth, labelDpi, report.Dpi);
                detail.MultiColumn.ColumnSpacing = XRConvert.Convert(labelInfo.HPitch - labelInfo.LabelWidth, labelDpi, report.Dpi);
                detail.MultiColumn.Layout = ColumnLayout.AcrossThenDown;
            }
            return report;
        }
        private ReportModel GetDefaultLabelReportMode(CustomLabelReportModel model) {
            XtraReportModel reportModel = new XtraReportModel();
            reportModel.ReportType = ReportType.Label;
            var customLblInfo = new CustomLabelInformation();
            customLblInfo.Height = model.LabelHeight;
            customLblInfo.Width = model.LabelWidth;
            customLblInfo.TopMargin = model.TopMargin;
            customLblInfo.BottomMargin = model.BottomMargin;
            customLblInfo.LeftMargin = model.LeftMargin;
            customLblInfo.RightMargin = model.RightMargin;
            customLblInfo.VerticalPitch = model.VerticalPitch;
            customLblInfo.HorizontalPitch = model.HorizontalPitch;
            customLblInfo.PaperKindDataId = model.PaperKindID;
            customLblInfo.Unit = model.MeasurementUnit;
            reportModel.CustomLabelInformation = customLblInfo;
            return reportModel;
        }
        private static PaperKindList InitPaperKindList(CustomLabelReportModel model, XtraReport report) {
            var paperKindList = new PaperKindList(report.Dpi);
            var labelRepository = new LabelProductRepositoryFactory().Create();
            var paperKindItems = labelRepository
                .GetSortedPaperKinds()
                .Select(ConvertToPaperKindItem);
            foreach (var item in paperKindItems) {
                paperKindList.Add(item);
            }
            paperKindList.CurrentID = model.PaperKindID;
            return paperKindList;
        }
        static PaperKindItem ConvertToPaperKindItem(PaperKindData paperKindData) {
            float dpi = DevExpress.XtraPrinting.GraphicsDpi.UnitToDpi(paperKindData.Unit);
            var paperKind = (PaperKind)paperKindData.EnumId;
            SizeF size = paperKind == PaperKind.Custom
                ? new SizeF(paperKindData.Width, paperKindData.Height)
                : PageSizeInfo.GetPageSizeF(paperKind, dpi, PageSizeInfo.DefaultSize);
            return new PaperKindItem(paperKindData.Name, size, paperKindData.Id, paperKind, dpi, paperKindData.IsRollPaper);
        }
    }
}
